ml-prims/src/linalg/coalesced_reduction.h

```diff
@@ -23,28 +23,38 @@
 namespace MLCommon {
 namespace LinAlg {
 
+
+template <typename Type, typename IdxType>
+struct MainNop {
+  HDI Type operator()(Type in, IdxType i) { return in; }
+};
+
+template <typename Type>
+struct Sum {
+  HDI Type operator()(Type a, Type b) { return a + b; }
+};
+
+
 // Kernel (based on norm.h) to perform reductions along the coalesced dimension
 // of the matrix, i.e. reduce along rows for row major or reduce along columns
 // for column major layout. Kernel does an inplace reduction adding to original
 // values of dots.
-template <typename Type, int TPB,
-          typename MainLambda = Nop<Type>, 
-          typename ReduceLambda = Sum<Type>,
-          typename FinalLambda = Nop<Type>>
-__global__ void coalescedReductionKernel(Type *dots, const Type *data, int D, int N, Type init, 
-                                         MainLambda main_op = Nop<Type>(),
-                                         ReduceLambda reduce_op = Sum<Type>(),
-                                         FinalLambda final_op = Nop<Type>(),
+template <typename InType, typename OutType, typename IdxType, int TPB,
+          typename MainLambda, typename ReduceLambda, typename FinalLambda>
+__global__ void coalescedReductionKernel(OutType *dots, const InType *data, int D, int N, OutType init, 
+                                         MainLambda main_op,
+                                         ReduceLambda reduce_op,
+                                         FinalLambda final_op,
                                          bool inplace = false) {
-  typedef cub::BlockReduce<Type, TPB> BlockReduce;
+  typedef cub::BlockReduce<OutType, TPB> BlockReduce;
   __shared__ typename BlockReduce::TempStorage temp_storage;
-  Type thread_data = Type(init);
-  int rowStart = blockIdx.x * D;
-  for (int i = threadIdx.x; i < D; i += TPB) {
-    int idx = rowStart + i;
-    thread_data = reduce_op( thread_data, main_op(data[idx]) );
+  OutType thread_data = init;
+  IdxType rowStart = blockIdx.x * D;
+  for (IdxType i = threadIdx.x; i < D; i += TPB) {
+    IdxType idx = rowStart + i;
+    thread_data = reduce_op( thread_data, main_op(data[idx], i) );
```

Comment:
Is the idea here to use this for argmin/argmax? In that case, will main_op and reduce_op return a tuple of some kind?

Reply:
@LeviBarnes Yes. The idea is that for argmin/argmax, the caller has to setup a struct of the following kind: `template <typename ValType, typename IdxType> struct pair { ValType val; IdxType idx; };`

Reply:
And also implement corresponding "reduction" operators on this struct.